coresimd/ppsv/codegen/masks_reductions.rs

```diff
@@ -0,0 +1,603 @@
+macro_rules! arm_64_x8_v7_neon_impl {
+    ($id:ident, $vpmin:ident, $vpmax:ident) => {
+        impl All for $id {
+            #[inline]
+            #[target_feature(enable = "v7,neon")]
+            unsafe fn all(self) -> bool {
+                use ::coresimd::arch::arm::$vpmin;
+                use ::mem::transmute;
+                // tmp = pmin(
+                //     (a, b, c, d, e, f, g, h),
+                //     (-, -, -, -, -, -, -, -)
+                // ) => (a, c, e, g, -, -, -, -)
+                let tmp = $vpmin(transmute(self), ::mem::uninitialized());
+                // tmp = pmin(
+                //     (a, c, e, g, -, -, -, -),
+                //     (-, -, -, -, -, -, -, -)
+                // ) => (c, g, -, -, -, -, -, -)
+                let tmp = $vpmin(tmp, ::mem::uninitialized());
+                // tmp = pmin(
+                //     (c, g, -, -, -, -, -, -),
+                //     (-, -, -, -, -, -, -, -)
+                // ) => (g, -, -, -, -, -, -, -).0 => g
+                let tmp: $id = transmute($vpmin(tmp, ::mem::uninitialized()));
+                tmp.extract(0)
+            }
+        }
+        impl Any for $id {
+            #[inline]
+            #[target_feature(enable = "v7,neon")]
+            unsafe fn any(self) -> bool {
+                use ::coresimd::arch::arm::$vpmax;
+                use ::mem::transmute;
+                // tmp = pmax(
+                //     (a, b, c, d, e, f, g, h),
+                //     (-, -, -, -, -, -, -, -)
+                // ) => (a, c, e, g, -, -, -, -)
+                let tmp = $vpmax(transmute(self), ::mem::uninitialized());
+                // tmp = pmax(
+                //     (a, c, e, g, -, -, -, -),
+                //     (-, -, -, -, -, -, -, -)
+                // ) => (c, g, -, -, -, -, -, -)
+                let tmp = $vpmax(tmp, ::mem::uninitialized());
+                // tmp = pmax(
+                //     (c, g, -, -, -, -, -, -),
+                //     (-, -, -, -, -, -, -, -)
+                // ) => (g, -, -, -, -, -, -, -).0 => g
+                let tmp: $id = transmute($vpmax(tmp, ::mem::uninitialized()));
+                tmp.extract(0)
+            }
+        }
+    }
+}
+
+
+// Implementation for ARM + v7 + NEON using vpmin and vpmax (folding
+// minimum/maximum of adjacent pairs) for 64-bit or 128-bit wide vectors with
+// more than two elements.
+#[cfg(all(target_arch = "arm", target_feature = "v7", target_feature = "neon"))]
+macro_rules! arm_128_v7_neon_impl {
+    ($id:ident, $half:ident, $vpmin:ident, $vpmax:ident) => {
+        impl All for $id {
+            #[inline]
+            #[target_feature(enable = "v7,neon")]
+            unsafe fn all(self) -> bool {
+                use ::coresimd::arch::arm::$vpmin;
+                use ::mem::transmute;
+                union U {
+                    halves: ($half, $half),
+                    vec: $id
+                }
+                let halves = U { vec: self }.halves;
```

Comment:
Is this a valid way to split a vector into its halves? It seems that clang uses shuffles for this purpose.

With this implementation, I see a bunch of functions whose core follows this pattern fail unit tests:

```rust
        #[inline(always)]
        pub fn simd_is_ascii(s: u8x16) -> bool {
            let above_ascii = u8x16::splat(0x80);
            s.lt(above_ascii).all()
        }
```

Reply:
> Is this a valid way to split a vector into its halves?

It is a valid way to do this. The RFC specifies that the layout of vectors is the same as arrays of the same element type and the same number of lanes, e.g., `f32x2` has the same layout as `[f32; 2]`. Since the layout of arrays is the same as the layout of tuples of the same element type and same number of element, e.g., `(f32, f32)` in the previous example, then yes, this is valid. I chose a tuple over an array here because I did not want to worry about doing `.get_unchecked`.

> With this implementation, I see a bunch of functions whose core follows this pattern fail unit tests:

On which arch? Can you post a minimal example that reproduces it here?

Reply:
> On which arch? Can you post a minimal example that reproduces it here?

ARMv7 with `-C target_feature=+neon`. I'll try to minimize it.

Reply:
Does this work on the other targets like `x86` and `x86_64` ?

Reply:
Yes. I tested that the pattern works on x86_64 by temporarily replacing the x86_64-specific code with the portable code.

Reply:
Sorry about misattributing the bug. The bug was a misplaced parenthesis in my conditional compilation directives, which caused ARMv7+NEON to use a stride size constant intended for ARMv7 ALU code.

So: Not a blocker for getting this PR merged.